lib/src/facts/aix/operating_system_resolver.cc

```diff
@@ -1,43 +1,84 @@
 
+// This routine's meant to be a general utility function that replicates the behavior of
+// lsattr -El <object> -a <field>. Although it's only used to get the modelname of the
+// sys0 device, we still would like to have it here in case we ever need to separate it
+// out to an AIX utils file to query other attributes. Part of what lsattr does is check
+// PdAt if we don't have an entry for the object's attribute in CdAt, even though it is very
+// unlikely that sys0.modelname will not have a CuAt entry. That is why we have the extra code
+// in here.
 static string getattr(string object, string field)
 {
-    string result;
-
-    execution::each_line(
-        "/usr/sbin/lsattr", {"-El", object, "-a", field},
-        [&](string& line) {
-            if (!line.empty()) {
-                vector<string> tokens;
-                boost::split(tokens, line, boost::is_space(), boost::token_compress_on);
-                if (tokens.size() < 2) {
-                    return true;
-                }
-                result = tokens[1];
-                return false;
-            }
-            return true;
-        },
-        nullptr,
-        0);
-
-    if (result == "") {
-        LOG_WARNING("Could not get a value from lsattr -El {1} -a {2}", object, field);
+    // High-level logic here is:
+    //   * Check if there's an entry for our object's field attribute in CuAt (the device-specific
+    //   attribute entry).
+    //
+    //   * Else, check for the field attribute's default value in PdAt. We do this by first
+    //   figuring out the PdDv type from the CuDv entry for object, then using our PdDv type
+    //   to query the field's default value in PdAt.
+    string query = (boost::format("name = %1% AND attribute = %2%") % object % field).str();
+    auto cdat_query = odm_class<CuAt>::open("CuAt").query(query);
+
+    // This is a more verbose way of saying that we only expect our query to have one element
+    auto cdat_ref = cdat_query.begin();
+    // TODO: Uncomment after testing!
+//    if (cdat_ref != cdat_query.end()) {
+//      auto value = string((*cdat_ref).value);
+//      if (value.empty()) {
+//        LOG_DEBUG("Could not get a value from the ODM for {1}'s '{2}' attribute.", object, field);
+//      }
+//      return value;
+//    }
+
+    // Get the PdDv type from the CuDv entry
+    query = (boost::format("name = %1%") % object).str();
+
+    auto cudv_query = odm_class<CuDv>::open("CuDv").query(query);
+    auto cudv_ref = cudv_query.begin();
+    if (cudv_ref == cudv_query.end()) {
+      LOG_DEBUG("Could not get a value from the ODM for {1}'s '{2}' attribute: There is no CuDv entry for {1}.", object, field);
+      return "";
+    }
+    auto cudv_obj = (*cudv_ref);
+
+    // TODO: Fill this in after figuring out PdDvLn
+    auto pddv_type = (*cudv_ref).PdDvLn;
+
+    LOG_DEBUG("NAME = {1}", cudv_obj.name); 
+    LOG_DEBUG("DDINS = {1}", cudv_obj.ddins); 
+    LOG_DEBUG("PDDVLN = {1}", cudv_obj.PdDvLn);
```

Comment:
@branan Unfortunately, it looks like `odm_get_first` and `odm_get_next` do not fill in the structure pointed to by `PdDvLn`. Based on what I have here, `cudv_obj.PdDvLn` prints 0, which is the null pointer. I looked elsewhere in the codebase, and we never directly reference `PdDvLn` like is done here. We only use it as a query parameter e.g. here: https://github.com/puppetlabs/facter/blob/master/lib/src/facts/aix/filesystem_resolver.cc#L87.

I'm not sure what I'm doing wrong here -- `cudv_obj.PdDvLn` seems to be the right type b/c I got a compile error earlier saying that it was of type `PdDv* const`. 

IBM's docs indicate that we could get things working with https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.basetrf1/odm_get_list.htm, which lets you specify the link depth. But this option would require some changes into the internal ODM api.

Reply:
Worst case, we could have this routine be something like `getattr(object, type, field)` where the type would be the uniquetype (or getattr(object, field, type = "") so we can have the option of skipping a check for PdAtr).

Reply:
Actually a cleaner approach might be to just create a separate iterator class that uses the struct array returned from `odm_get_list` as the underlying container. We could have the interface for open be something like
```
odm_class<T>::open(class_name, expand_link_descriptors = NO)
```

and conditionally create either our current iterator or the `get_list` iterator. It would probably not be a good idea to use `get_list` off the getgo b/c that returns everything at once, which seems kind of bad (if I'm reading the documentation right).

I can look into this if we want to look at this approach.

Reply:
I just poked `/usr/include/sys/cfgodm.h` to see how the struct actually stores its link. There are three values associated with it - the struct pointer is the `PdDv` field, but there is also a `PdDvLn_Lvalue` field that is a `char[]` - that may be the string form of the link?